api/action/index.js

```diff
@@ -0,0 +1,120 @@
+const EMOJI = [
+  'confetti',
+  'suit',
+  'bow',
+  'dog',
+  'skull',
+  'yoyo',
+  'cat',
+];
+
+let buttons = document.querySelectorAll('button');
+
+// The action API does not expose a way to read the action's current enabled/disabled state, so we
+// have to track it ourselves.
+// Relevant feature request: https://bugs.chromium.org/p/chromium/issues/detail?id=1189295
+let actionEnabled = true;
+let showToggleState = document.getElementById('show-toggle-state');
+document.getElementById('toggle-state-button').addEventListener('click', (_event) => {
+  if (actionEnabled) {
+    chrome.action.disable();
+  } else {
+    chrome.action.enable();
+  }
+  actionEnabled = !actionEnabled;
+});
+
+document.getElementById('set-popup-button').addEventListener('click', () => {
+  let popup = document.getElementById('popup-options').value;
+  chrome.action.setPopup({ popup });
+});
+
+let getPopupButton = document.getElementById('get-popup-button');
+getPopupButton.addEventListener('click', () => {
+  chrome.action.getPopup({}, (result) => {
+    document.getElementById('current-popup-value').value = result;
+  });
+});
+
+// Badge Text
+
+document.getElementById('set-badge-text-button').addEventListener('click', () => {
+  let text = document.getElementById('badge-text-input').value;
+  chrome.action.setBadgeText({ text });
+});
+
+document
+  .getElementById('get-badge-text-button')
+  .addEventListener('click', () => {
+    chrome.action.getBadgeText({}, (result) => {
+      document.getElementById('current-badge-text').value = result;
+    });
+  }
+  )
+
+// Badge background color
+
+document.getElementById('set-badge-background-color-button').addEventListener('click', () => {
+  // To show off this method, we must first make sure that there is a badge value
+
+  chrome.action.setBadgeText({ text: 'hi :)' });
+
+  let color = new Array(4);
+  for (let i = 0; i < color.length; i++) {
+    color[i] = Math.floor(Math.random() * 256);
```

Comment:
Does this actually respect opacity?

Reply:
It doesn't. An alpha value of 0 will set the badge color to the default value, any other value will set the badge to a fully opaque version of the specified color.

https://bugs.chromium.org/p/chromium/issues/detail?id=1184905

Reply:
It feels confusing to set the alpha channel to a random value, then? Perhaps you should have a:

```js
let color;
if (Math.random() < 0.5) {
  // show default color
  color = [0, 0, 0, 0];  // zero opacity uses default color
} else {
  color = new Array(3).map(() => Math.floor(Math.random * 255));
  color.push(255);  // if non-zero, this will be fully opaque
}
chrome.action.setBadgeBackgroundColor({ color });
```

Reply:
I like this suggestion. I'll integrate this today.

Reply:
FYI, `new Array(3).map()` does not work as expected because new Array produces a holy array and `map()` skips holes. A couple alternative oneliners are

```js
// generic
new Array(n).fill().map(…);
[...new Array(n)].map(…);
Array.from(new Array(n).values()).map(…);

// specific
[0,0,0].map(…)
```

In this case I think I'll go with the simplest approach and just use a literal.